### DISCUSSION TITLE
RMT - 2 Channels syncron?

### DISCUSSION THREAD

**Discussion by zmaier24** [micropython/micropython#9298] [General]
Hello,

i have a "problem" with RMT. I want to use 2 channels just for sending a sequence. For test reasons the sequence is just 1,0,1,0,....
And the 2 channels should start synchron. I am using Micropyhton 1.18 on a Wemos Mini ESP32.

Now, the problem
1) Both channels are not starting synchron and
2) even the sequence is the same, the output is not the same, it is inverted?

![oszi](https://user-images.githubusercontent.com/63282373/189849968-db7974b5-98a7-4ff3-9f30-7fe29c8f53c3.jpg)

Any ideas why?

Here the code:

```import esp32
from machine import Pin
try:
    r = esp32.RMT(0, pin=Pin(21), clock_div=80)
    r1 = esp32.RMT(1, pin=Pin(22), clock_div=80)
    r.loop(1)
    r1.loop(1)

    r.write_pulses((100, 100), 1)
    r1.write_pulses((100, 100), 1)
except:
    r.deinit()
    r1.deinit()

**Comment by peterhinch** [comment]
It would be interesting to try this with single-shot sequences.

**Reply by zmaier24** [reply]
Unfortunately  i killed the device, i have to wait till the new ones are arriving :(

**Comment by jonathanhogg** [comment]
It's unlikely to be inverted. It's more likely to be ~180° out of phase. A non-symmetrical sequence (e.g., `(120, 80)`) would confirm this.

I can't think of any way to start two RMT channels in sync. The best you would get is starting them *very close to one another*, but that would require large changes in the MicroPython RMT implementation: in the current design there is a not-insignificant amount of setup being done when you call `.write_pulses()` before the hardware is triggered.

**Comment by zmaier24** [comment]
You are right, it is not inverted, there is just out of phase, since there is a time delay. In my config, the time delay is exactly as big as that the signal looks like it is inverted :) 
I played with a `time.sleep_us` between the two `.write_pulses(). to sync the two streams, but this was also not really working, because the waiting time needed was not constant (After a reboot).  Which is strange, i thought the programm is deterministic? 
I need between 135 and 145 µs time delay.

```
r.write_pulses((100,50),1)# 100 = 100µs
time.sleep_us(140)
r1.write_pulses((100,50),1)
```
Any idea how i could generate a sequence and the exact inverted sequence (this is actually the usecase for me).

**Reply by robert-hh** [reply]
At the moment, the only reliable method is using an external inverter. If a high impedance/moderate rise time can be accepted, then it could be as simple as a single 2n7000 transistor and a resistor.

**Reply by zmaier24** [reply]
Thx for the hint, i made a simple Switch with a BC337, it works perfect :)

**Comment by mattytrentini** [comment]
A Sync Manager allows multiple channels to be started simultaneously, but such configuration isn't yet exposed through MicroPython. See [Multiple Channels Simultaneous Transmission](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html#multiple-channels-simultaneous-transmission). 

I think if we put our heads together we can come up with something to provide this functionality!

**Reply by russhughes** [reply]
That looks interesting, it would require using a newer version of esp-id.
